### ISSUE TITLE
KNX Implimentation

### ISSUE THREAD

**Issue by Alex0176** [Aircoookie/WLED#2568]
**Is your feature request related to a problem? Please describe.**
The featurerequest is primary for use WLED via KNX Bus or better the Presets for KNX Group Addresses. 

**Describe the solution you'd like**
There is a git project for use KNX and can be used for set the KNX Address and send and receive data from/to KNX Bus with a Bus ,,koppler". In the Video you see whats needed for KNX and the wireing:
https://www.youtube.com/watch?v=KtBNI-xF23w

the KNX lib can be downloaded here:
https://knx-user-forum.de/filedata/fetch?id=1549797

- So you can select presets via Group Adresses
- Send % values via KNX (for example for garage doors or water filled or temp 

**Describe alternatives you've considered**
In WLED is needed to:
- Implement the KNX Lib.
- Add a enable KNX with the KNX Adress (this must be the same like a Dummy device in ETS)
- Add a possibility for Group Addresses for sending and receiving the values. 

Values:
Send to KNX: actual preset, on/off WLED (for Powersupply)
Receive from KNX: set preset, set % value (for special effects)

**Additional context**

Thank you for your ideas for making WLED better!

**Comment by Aircoookie** [comment]
This is a good idea and would be best implemented as a WLED usermod by a knowledgable user with a KNX system :)

**Comment by Alex0176** [comment]
i think the parts todo: ( i can read C/C++, but i have no knowlage in this) 
+ add knx lib 
+ add on init file (i dont know where to find) if knx is enabled set the correct physical adress to the knx lib.

if (knx ==true) {
    KnxTpUart knx(&Serial1, "x.x.x"); // "x.x.x" should be the stored physical adress in cfg.cpp
    Serial1.begin(19200, SERIAL_8E1)
    knx.uartReset();
}
+ add the information into the preset.cpp (for save/load the settings)
+ add enable knx and physical adress to cfg.cpp
+ modify the html templates 
+ make a thread loop for the knx part and fire up events on the correct lines of code to knx, if its enabled. 
for example in present.cpp (hopefully thats the correct point) 
... applyPresent(.....)
if(knx ==true && presentGroupAdd!=null){

bool success = knx.groupWriteBool("x/x/x", index);
.....

**Comment by Alex0176** [comment]
Today i had 2 hours time and check out the usermod v2 example and i make a few things, but i dont know how i can add config fields to define the group address directly at the presents (for now i make a possibility of 50 presets hardcorded and i dont know how to compile the staff. My atom IDE is sucking very at the autocomplete and dont like the code from wlan in my configuration. 

but i only start atom and check out the wled code and make a modification from that as base of my usermod. 

Can anyone compile it for me or help me with my questions?
[KNX-Presents.zip](https://github.com/Aircoookie/WLED/files/8673492/KNX-Presents.zip)

**Comment by Alex0176** [comment]
i got now a link error, but i'am not able to fix this. i think there is a issue with casting the serial port to a stream. 

Can anyone help me in this case? 

```
/../../../xtensa-lx106-elf/bin/ld.exe: .pio\build\d1_mini\src\usermods_list.cpp.o: in function `_GLOBAL__sub_I_physicalAddress':
usermods_list.cpp:(.text.startup._GLOBAL__sub_I_physicalAddress+0x2b): undefined reference to `KnxTpUart::KnxTpUart(Stream*, String)'
collect2.exe: error: ld returned 1 exit status
*** [.pio\build\d1_mini\firmware.elf] Error 1
```
[KNX-Presents-a0.0.2.zip](https://github.com/Aircoookie/WLED/files/8842591/KNX-Presents-a0.0.2.zip)

**Comment by blazoncek** [comment]
Please use WLED [forum](https://wled.discourse.group) or Discord for help and support questions.

**Comment by Alex0176** [comment]
> @Alex0176 Have you made any progress one this?

i dont know why it does not compile ... and noone like to help me in this case.

**Comment by softhack007** [comment]
> Today i had 2 hours time and check out the usermod v2 example and i make a few things

Welcome to the world of WLED development :-)
Not to spread frustration, but I'd encourage you to spend more time on the task. 2 hours is surely to little time to even understanding the basic.

----

> #include <KnxTpUart.h>
> ^
> compilation terminated.
> undefined reference to `KnxTpUart::KnxTpUart(Stream*, String)'
collect2.exe: error: ld returned 1 exit status

This looks like platformIO cannot find a library that provides KnxTpUart.h. which library do you want to use? 
This library looks similar to what is missing: https://github.com/majuss/KnxTpUart


----
> add the information into the preset.cpp (for save/load the settings)
> add enable knx and physical adress to cfg.cpp
> modify the html templates


Actually you don't need to fiddle with set.cpp + cfg.cpp and  html templates, if you  want to make you own usermod settings. 

This can be done inside the usermod using the standard JSON API.
https://github.com/Aircoookie/WLED/blob/74156b7ed8a1449ed459baeb7934debb53a19665/usermods/EXAMPLE_v2/usermod_v2_example.h#L187-L197

If you need a full example, take a look at the boblight usermod 
https://github.com/Aircoookie/WLED/blob/74156b7ed8a1449ed459baeb7934debb53a19665/usermods/boblight/boblight.h#L307-L351

**Comment by Alex0176** [comment]
> KnxTpUart.h

Hi, the libs are included in the ZIP File, and i dont get this kind of compile error. I get the error `/../../../xtensa-lx106-elf/bin/ld.exe: .pio\build\d1_mini\src\usermods_list.cpp.o: in function `_GLOBAL__sub_I_physicalAddress':
usermods_list.cpp:(.text.startup._GLOBAL__sub_I_physicalAddress+0x2b): undefined reference to `KnxTpUart::KnxTpUart(Stream*, String)'
collect2.exe: error: ld returned 1 exit status
*** [.pio\build\d1_mini\firmware.elf] Error 1`

So its not a libary problem, more a linker or add cpp problem. i dont know - i am not a c programmer, so its hard for me to figure out whats exacly the problem right now.

**Comment by softhack007** [comment]
> So its not a libary problem, more a linker or add cpp problem. i dont know - i am not a c programmer,

Actually it is a library problem, as the linker does not find the library where the class KnxTpUart is implemented. 

This typically happens in platformIO if you manage to feed it with a header file, but did not use the proper way of importing an external library. The usual way to do so, is by using `lib_deps` and pointing to a github repository that provides the library.

**Comment by Alex0176** [comment]
i dont know how i can configure this - so i hope anyone other with C knowlege can work on it. i can tast and help with the logic

**Comment by PaoloTK** [comment]
> i dont know how i can configure this - so i hope anyone other with C knowlege can work on it. i can tast and help with the logic

I'm working on a KNX TP integration myself. I have switch groups working but that's it for now. I was wondering, other than listening and writing to groups that you would expect from a light (on/off, brightness, effect, etc.) is there any other feature you think makes sense to have?

**Comment by Alex0176** [comment]
i think the power on/off for the powersully will be a good idea to have. So you can power the ESP via KNX and switch the powersupply via KNX group. The other effects you find on the typical switch solution, like MDT have on his glas panals. 
https://www.youtube.com/watch?v=cx85Y5eTSHQ

**Comment by PaoloTK** [comment]
> i think the power on/off for the powersully will be a good idea to have. So you can power the ESP via KNX and switch the powersupply via KNX group. The other effects you find on the typical switch solution, like MDT have on his glas panals. https://www.youtube.com/watch?v=cx85Y5eTSHQ
> 
> did you have already a usemod or custom firmware?

The power supply should switch on automatically when the light is turned on in WLED, why would you add the relay to a separate switch group?

You can find the usermod code in my repos, WLED/KNX. But it's too early to submit it, needs a lot more work done.

**Comment by Alex0176** [comment]
yes, but if you make this with ESP32 than you have to switch the 230V relay, but if you have this already on KNX, you can use this switch. This make sense i think.

**Comment by PaoloTK** [comment]
> yes, but if you make this with ESP32 than you have to switch the 230V relay, but if you have this already on KNX, you can use this switch. This make sense i think.

I have already implemented a light feedback group. If you have an actuator that controls the light power supply, add it's switch group to the light feedback group. When you turn the light on in WLED it will send a "TRUE" to the group, turning on the actuator.

Is this what you were looking for?

**Comment by PaoloTK** [comment]
@Alex0176 I have made some further progress and added basic instructions on getting my usermod up and running. I want to stress this is super early and there will be bugs, but if you want to try it I've got basic functionality (on/off, absolute and relative dimming) working. You can find the instructions and code here, but you will have to custom compile. I haven't tested the usermod with free and 2-level group style, so it won't probably work unless you use 3-level style.

https://github.com/PaoloTK/WLED/tree/KNX/usermods/KNX

If you have any questions or have troubles getting this up and running, let me know.

**Comment by KlausHans** [comment]
Very interesting idea! Tasmota has this feature already, the integration in Wled would be perfect.
@PaoloTK I would like to test but i only have a knx interface, no router. I think i will change that in the future, but for now i can not test it.

**Comment by PaoloTK** [comment]
> Very interesting idea! Tasmota has this feature already, the integration in Wled would be perfect. @PaoloTK I would like to test but i only have a knx interface, no router. I think i will change that in the future, but for now i can not test it. Edit: i just saw that you dont use Knx Ip.

The interface will work fine too, the WLED devices communicate over twisted pair.

**Comment by michael-blue2** [comment]
> @Alex0176 I have made some further progress and added basic instructions on getting my usermod up and running. I want to stress this is super early and there will be bugs, but if you want to try it I've got basic functionality (on/off, absolute and relative dimming) working. You can find the instructions and code here, but you will have to custom compile. I haven't tested the usermod with free and 2-level group style, so it won't probably work unless you use 3-level style.
> 
> https://github.com/PaoloTK/WLED/tree/KNX/usermods/KNX
> 
> If you have any questions or have troubles getting this up and running, let me know.

Hey @PaoloTK I have been trying to compile your code to test it. 
But I do always get an error that (PackageException: Package version 2.2.0+sha.a68a18c doesn't satisfy requirements ~2.0.7 based on PackageMetadata) 
Any chance you can share the compiled file (.bin file)? 
And have you made any process from the first time you uploaded your code? as its been 7 months.

**Comment by michael-blue2** [comment]
I got around that error and just compiled the bin file. I will start right now the test with the bus system. I will be using some abb bus couplers ( a bit over priced) but I already had them around.

**Comment by PaoloTK** [comment]
> > @Alex0176 I have made some further progress and added basic instructions on getting my usermod up and running. I want to stress this is super early and there will be bugs, but if you want to try it I've got basic functionality (on/off, absolute and relative dimming) working. You can find the instructions and code here, but you will have to custom compile. I haven't tested the usermod with free and 2-level group style, so it won't probably work unless you use 3-level style.
> > https://github.com/PaoloTK/WLED/tree/KNX/usermods/KNX
> > If you have any questions or have troubles getting this up and running, let me know.
> 
> Hey @PaoloTK I have been trying to compile your code to test it. But I do always get an error that (PackageException: Package version 2.2.0+sha.a68a18c doesn't satisfy requirements ~2.0.7 based on PackageMetadata) Any chance you can share the compiled file (.bin file)? And have you made any process from the first time you uploaded your code? as its been 7 months.

Hi, glad you got it compiled, let me know how it works. As for progress on the code, I got stuck due to performance issues when trying to support 600+ groups (all settings for each segment for max amount of segments), so I started a complete rewrite with a more efficient memory handling but got kinda stuck and been busy with work. Not sure if it's worth the effort since I doubt most people use that amount of segments.

**Comment by Alex0176** [comment]
i dont get it up and running right now - so i wait for a working image or something like this, where we can configure the knx stuff via webinterface. . . than i will add the knx stuff to my wled installations here.

**Comment by HoffmannOs** [comment]
Any News?

**Comment by PaoloTK** [comment]
> Any News?

Looking to start back development on the feature in the next few days. I think it makes sense to ship an MVP and worry about additional features later.

**Comment by softhack007** [comment]
> I think it makes sense to ship an MVP

@PaoloTK sorry I'm not a native speaker - what's an MVP, and what happens when you put it onto a ship?

**Comment by fdepalo** [comment]
Hey Paolo, i don't know but i imagine we speak the same language. I developed a new tpuart adapter that is working flawlessly with esphome and i just managed to compile your usermod. Do you need help ?

**Comment by PaoloTK** [comment]
> Hey Paolo, i don't know but i imagine we speak the same language. I developed a new tpuart adapter that is working flawlessly with esphome and i just managed to compile your usermod. Do you need help ?

Hey. :) 
I did start working on a new KNX library but it's a lot of work, so was thinking of sticking to the current one for the time being, even if it has some questionable decisions (e.g. only passing group addresses as String). I found the openKNX library to be extremely well written but its license is incompatible with KNX and it's only designed to be used with ETS (can't configure device from WLED GUI).

**Comment by michael-blue2** [comment]
Do we have any update on this? 
I have been waiting for some months now to test this.  Just finished my first knx project which I haven't yet installed my wled devices.
My first idea was to connect them through home assistant and expose them from there to knx keypads but I do prefer them directly to knx.

**Comment by marcel151** [comment]
Any news? I am planning lighting for our new home and also want to realise a few RGBCCT strips (e.g. garage, stairs). I think WLED is _the_ solution for adressable LED strips. Also nowadays there are components that are somewhat "professional" and not the DIY solutions as before, for example this POE module for DIN rail mount installation:
https://shop.codm.de/en/automation/wled/wled-lan-controller-poe-din-rail-range-out

Unfortunately WLED is lacking is KNX support, which more and more homes in Europe (especially in Germany) are equipped with. KNX is going from a former more industrial automation standard to an also Smart Home automation solution. Since it's a well-used standard, you don't have to worry about end of support in the future. That's why it is the solution we are going for in our home.

I think KNX IP support would be ideal for WLED, since you don't need new hardware for it.

I am not a software developer. But I am able to read, understand and alter existing code and happy to assist in testing (when our home is finished).

**Comment by u20p17** [comment]
@PaoloTK: thanks for all your effort! 👍 any news on this topic? would be great if it would also work via knx-net in the future 👌

edit: today I checked a bit existing KNX integrations for ESP32 and I found the one from @thelsing (https://github.com/thelsing/knx) which looks very nice. Still we would face the same license issue, as it is licensed under[GPL-3.0 license](https://github.com/thelsing/knx#GPL-3.0-1-ov-file) and WLED under [EUPL-1.2 license](https://github.com/wled/WLED#EUPL-1.2-1-ov-file)...

**Comment by u20p17** [comment]
I did more research on this topic and found another lib: https://github.com/envy/esp-knx-ip This lib has an [MIT license](https://github.com/envy/esp-knx-ip#MIT-1-ov-file) and could be used within a usermod in WLED without any license issues.  Of course it is not as advanced as the lib from @thelsing, so you do not have 'full KNX device behavior' (ETS integration, datapoints, twisted pair bus), but I think this would be already a big step in the right direction...

**Comment by u20p17** [comment]
Hi all,
I started to implement a KNX_IP usermod v2 to wled v1.15.1 (https://github.com/u20p17/WLED/tree/release/v0.15.1). The whole integration is still untested in a real KNX system as I am still in the planning phase of my house :P Therefore it would be great if somebody of you could test it and give feedback/make commits with bugfixes. (@Alex0176 could you give it a go? - i can also send you a precompiled firmware for an esp32 with knx-ip usermod)

(I took the https://github.com/envy/esp-knx-ip  library as reference (removed webserver and some stuff) as it has the MIT license...)

**Comment by dennnnis99** [comment]
Hi @u20p17, i would love to test the firmware with real hardware. I am currently renovating my complete flat and I would love to have the flexibility of wled and the stability of Knx combined. I plan to use these wled controllers (https://amzn.eu/d/2G8KanZ) because I already have two of those and they work great. They are based on esp32 so I assume they should be compatible with the firmware you have compiled.

**Comment by netmindz** [comment]
Can @u20p17 you create a branch and PR for your changes?

**Comment by fdepalo** [comment]
> Hi all,
> I started to implement a KNX_IP usermod v2 to wled v1.15.1 (https://github.com/u20p17/WLED/tree/release/v0.15.1). The whole integration is still untested in a real KNX system as I am still in the planning phase of my house :P Therefore it would be great if somebody of you could test it and give feedback/make commits with bugfixes. (@Alex0176 could you give it a go? - i can also send you a precompiled firmware for an esp32 with knx-ip usermod)
> 
> (I took the https://github.com/envy/esp-knx-ip  library as reference (removed webserver and some stuff) as it has the MIT license...)

hey @u20p17, if you are interested i developed a knx-tp board for esp32, i don't know how much work can be integrating it. Let me know 🤪

**Comment by u20p17** [comment]
@dennnnis99 : I uploaded the first beta version here: https://github.com/u20p17/WLED/releases/tag/v0.15.1_KNXIP_beta1 If you can try it would be great - I expect some bugs in there, as I couldn't test it up to now...

@netmindz : I will make a PR as soon as I get positive feedback from some KNX users that it is working fine.

@fdepalo: I will have to check how much work it might be... Still would be interessing for me as well. Are you in the knx-user-forum.de? Maybe we could write there via PM - I have the same name as on github (u20p17)?

**Comment by marcel151** [comment]
@u20p17 Thanks for your efforts trying to get KNX to WLED, hopes are high that we'll get a functional version in the future. I am happy to help when we finish our KNX powered home.
I am also interested in a KNX TP version. If there are efforts to integrate it maybe there is a use case to make KNX-TP ESP32 hardware (thanks @fdepalo). I think this would also be interesting for [@codmpm](https://github.com/codmpm), as he is also developing WLED hardware, which are high quality components.

**Comment by HoffmannOs** [comment]
Have Flash on a esp32Wroom to test it, in Usermods KNX Shows up, PA Can change to 0.0.12x, but GA dont can change to another Adress.... click Save and reboot, but only the 1/0/1 will repeat.

**Comment by u20p17** [comment]
@HoffmannOs : the JSON keys didn't match between save and load... here is a fixed version (Beta2): [WLED_0.15.1_ESP32_KNX.bin](https://github.com/u20p17/WLED/releases/download/v0.15.1_KNXIP_beta2/WLED_0.15.1_ESP32_KNX.bin)

**Comment by HoffmannOs** [comment]
> [@HoffmannOs](https://github.com/HoffmannOs) : the JSON keys didn't match between save and load... here is a fixed version (Beta2): [WLED_0.15.1_ESP32_KNX.bin](https://github.com/u20p17/WLED/releases/download/v0.15.1_KNXIP_beta2/WLED_0.15.1_ESP32_KNX.bin)

THX, have Test it, can now Save the GA, but not send or arrived on BUS, have it run with many Devices with Tasmota without any Problems, the Procedure is the Same.

**Comment by u20p17** [comment]
ok i will check it again... while testing I checked in the debug console if the messages are send and there it seemed that it is working... now i also checked it with wireshark, but i can not see messages there too... I will check the code again... maybe it is somehow related to the multicast and wifi? do you also have a esp32 with ethernet to test? (if so I could try to build an firmware with ethernet and you could test this too)

**Comment by HoffmannOs** [comment]
@u20p17 yes, wireshark dont show up. Have check too now.
No i only have Wlan Devices... no Ethernet.
I use it in Unifi and have integrated around 20 Tasmota devices, they all work smoothly. I had problems with a KNX IP router, but I was able to fix them.

**Comment by Alex0176** [comment]
thats good news - a TP was mutch better, because we can add a KNX hardware with a relay to switch on a power supply if we need it. and in the time between the ESP32 get power via Bus. 
That makes it easier and more power saving.

**Comment by michael-blue2** [comment]
I had the same idea when I was designing/wiring my house. I dont know if the TP intigration will ever happen, i have one knx relay on my 24v psu for the wled. and for now i am just using home assistant to change the colors. Unfortunately I do not have an ip gateway to test the new code but only an ip interface as i wasnt planning on having any KNX IP (as i have wired every single place in the house with a TP cable).  I am just happy to see some traffic here though :) worst case scenario i will just grap a knx gateway, just to be able to loose the home assistant in any client that only runs on knx.

**Comment by DanielWeeber** [comment]
I have not read every comment here but I do have a working KNX System and a WLED in place. If there is anything you would like me to test or even do a remotecontrol session with me or similar just hit me up. We can also switch to german ;)

**Comment by u20p17** [comment]
@HoffmannOs : can you please give it a try with beta3 ? https://github.com/u20p17/WLED/releases/tag/v0.15.1_KNXIP_beta3

With this version I was able to see (some) Multicast messages with wireshark. The ESP32 itself recieved all messages which were send by itself... maybe it is working now :P

**Comment by HoffmannOs** [comment]
@u20p17 no, not Working... i see the Led from Esp32 now blinking constant, wireshark i check tomorrow.

**Comment by u20p17** [comment]
ok I will check it again. i can see some of the messages in wireshark:

<img width="1024" height="857" alt="Image" src="https://github.com/user-attachments/assets/ea8ed1eb-27fe-4210-8367-bc508590f647" />

**Comment by u20p17** [comment]
beta4 is available ( https://github.com/u20p17/WLED/releases/tag/v0.15.1_KNXIP_beta4 ), where changing GAs and individual address is changing immediatly after pressing apply (so no need to restart the esp)... one more thing i figured out for capturing the multicast messages with wireshark: I have to run a small python script (joins to multicast group)  in parallel to wireshark to see the messages. according to KI it is related to the following reason:

> On most OSes, multicast packets from another host (your ESP32 on Wi-Fi) will not be forwarded to your capture interface unless the OS has joined the multicast group.
> 
> Ethernet drivers often drop packets for groups nobody on the host has subscribed to (to save CPU).
> 
> WinPcap/Npcap on Windows is especially strict: if the OS didn’t “IGMP join” 224.0.23.12, the NIC filters the packets before Wireshark can see them.


Here the script I am running (you will have to change the ip accordingly):

```
import socket, struct
MCAST_GRP = '224.0.23.12'
MCAST_PORT = 3671
LOCAL_IP = '192.168.99.242'   # <-- put your Wi-Fi adapter IPv4 here

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
sock.bind(('', MCAST_PORT))

mreq = struct.pack('=4s4s', socket.inet_aton(MCAST_GRP), socket.inet_aton(LOCAL_IP))
sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)

print(f'Joined {MCAST_GRP}:{MCAST_PORT} on {LOCAL_IP}, waiting for packets...')
while True:
    data, addr = sock.recvfrom(2048)
    print(f'from {addr}: {data.hex()}')
```

**Comment by HoffmannOs** [comment]
@u20p17  Sorry, dont know how to use the Script, I Use 2 Esp32 and 1 RP2040 Ip Routers with LAN/TP from Openknx Forum, these are Running without any Problems on Tasmota, in Wireshark dont see anything from these IP of My Esp32 Wled Beta1-4....
My Unifi Setup and the Switches have correkt IGMP Querier and Config.
At moment i use my WLED Controllers all over Iobroker, plan is these over my Presence Modul Applikation from OPEN KNX Steering. Sorry i can help you with the Script, tell me when New Beta to test.

**Comment by u20p17** [comment]
@HoffmannOs : Do you see the KNXIP messages in wireshark from the tasmota devices? If so, can you send me of these message? 

To run my script you would have to copy the code snippet in a .txt file, change the ip address to fit to you laptop/pc ip and save it. after that change the filename to 'test.py' and open the cmd window. navigate to the folder where the .py file is located and type 'python .\test.py' to run the script (of course you need python installed on the pc/laptop...). If I do this I can see the KNXIP messages from the Beta4 in wireshark on my system...

**Comment by u20p17** [comment]
beta 6 is out now: https://github.com/u20p17/WLED/releases/tag/v0.15.1_KNXIP_beta6. I found out that the cEMI message code was formatted wrong... 

A multicast message i see with my script when changing the global brightness to 95 in the wled gui is: 0610053000122900bce01164100202008025

Decode (byte-wise):

06 10 → KNXnet/IP proto 0x10
05 30 → Routing Indication
00 12 → total length = 18 bytes
29 00 → cEMI L_Data.ind, add.info=0 
BC E0 → CF1/CF2: std frame, group addr, hop=6 (sane defaults)
11 64 → source 1.1.100
10 02 → dest GA 1/0/2 (your brightness GA)
02 → cEMI “data length” = 2 (TPDU[1] + 1 data byte)
00 80 → APCI = GroupValueWrite
25 → payload = 0x25 = 37 (wled brightness is 0-255 --> scales 95/255 → ≈ 37 %... send in % as we are using DPT 5.001)

so i believe now the message should be formatted correctly. 

@HoffmannOs : can you test beta 6 again and if it still is not working send me some messages recorded with wireshark of the tasmota devices? (or use my python script to log the data from the tasmota devices...) Thanks!

**Comment by HoffmannOs** [comment]
@u20p17 Hey, today i was very busy.... i have test the beta 6, in ETS dont come what on Busmonitor, or the ESP32, tomorrow i can send you the messages from Tasmota with Wireshark. Yesterday have use the same Controller, flash Tasmota Knx, Enable and work directly after pressing save. 

Maybe take a look here https://github.com/OpenKNX

Thx for your intensive Work to get this feature on.

King Regards

**Comment by u20p17** [comment]
@HoffmannOs : I did some further tests where I flashed the ESP32 with Tasmota and introduced a PWM Output, which i mapped to GA 2/0/2. Then I logged the payload with the same python script I used also for the KNX_IP usermod:

the Tasmota payload looks like:
0610053000122900bce0116410020200802b

the payload of the usermod looks like:
0610053000122900bce01164100202008025

```
06 10 05 30 00 12   → KNXnet/IP Routing Indication (len=0x12 = 18)
29 00               → cEMI L_Data.ind, no add. info
BC E0               → ctrl (std frame, group addr, low prio)
11 64               → source IA = 1.1.100
10 02               → dest GA = 2/0/2
02                  → TPDU data len = 1 byte
00 80 xx            → GroupValueWrite, DPT 5.001 (Scaling 0..100 %), value = xx
```

What I can see the KNX telegram type is the same between both (only difference is the % output which I send (2B-->43%, 25--> 37%)

so for me it looks like it should work correctly... I will do the same for recieving (didn't have time up to now...) But if you make a brightness change on the WLED GUI you should see the message being send from the usermod...

**Comment by HoffmannOs** [comment]
Here are my Results from Tasmota,
hope thats correctly for you? 

<img width="1914" height="943" alt="Image" src="https://github.com/user-attachments/assets/c5e02a1f-e7f1-4d1b-86aa-91b49514fcfb" />
<img width="1596" height="608" alt="Image" src="https://github.com/user-attachments/assets/e408db06-8a98-4d24-901d-fbe6b1c97c0d" />
<img width="603" height="862" alt="Image" src="https://github.com/user-attachments/assets/eb9e8644-c39c-48c9-9a34-228870a26b4e" />

Dont can use the group 2/0/xx.... can you use a beta with my Group adresses 6/x/x to test, when i change these in WLED maybe dont save ? 

Dont have see one Message from Wled Controller in Wireshark, or Bus Monitor.

**Comment by u20p17** [comment]
Thanks for testing again! What I can see from your wireshark snippet is that the GA 6/0/3 payload looks like this with Tasmota:

06 10 05 30 00 11 29 00 BC E0 00 7C 30 03 01 00 80

I changed the On/Off GA out also to 6/0/3 in the usermod and checked the output -->

06 10 05 30 00 11 29 00 BC E0 00 7C 30 03 01 00 80

So how I generate the payload seems to be fine. What I saw is that you have the same GA for IN and OUT, which might cause an issue in the older betas. In the latest one (Beta8: https://github.com/u20p17/WLED/releases/tag/v0.15.1_KNXIP_beta8 ) I change a few things:
1.) If the WLED esp sends and recieve a message to the same GA it now autodetects if it was send on the Bus by itself and ignores it.
2.) I also changed the logic behind the rejoin of the multicast if you change GAs or any other KNX related input fields...
3.) I also added the same 'strategy' as Tasmota has when Communication enhancement is ticked ( resend every multicast messafe several times ( can be configured - standard 3times)


@HoffmannOs : would be great if you could try again the latest beta and if you still can not see any messages in wireshark /ETS it would be great if you change the GA _in to different GAs as the GA_out (just for an test). After you saved this try to changeon/off via the WLED GUI. If this still does not help please go to https://tasmota.github.io/install/ open the connection to the esp and then press Log&Console. Then you should see the log messages and when changing values in the WLED Gui you should see there if messages get send. Please send me the log to see if I can find any issue there. Thanks!

**Comment by marcel151** [comment]
I just managed to get a Demo running with KNX Virtual. I think this would be also suitable for you @u20p17 since you also don't have a physical KNX installation for now. Both KNX Virtual and ETS Demo (up to 5 Devices) are free.

https://www.knx.org/knx-de/fuer-fachleute/erste-schritte/knx-virtual/
For Multicast communication to WLED you have to configure the router in KNX virtual, that's it:
https://support.knx.org/hc/de/articles/4502245075730-Router-Verbindung-mit-einer-physischen-Installation

I did a test and it seems to work fine. I can switch WLED on from virtual KNX switch and feedback is also working.

Attached is my WLED KNX config and group monitor log. As you can see the bus recognises when switching on WLED and also I can turn WLED on from KNX Virtual. Feedback works since Channel 3 is red, which means on.
Correct me if I'm wrong, you also need to have a KNX IP Router, not an IP interface in a physical installation, since [esp-knx-ip](https://github.com/envy/esp-knx-ip) uses Multicast addresses. Worth mentioning I think.

Just tell me if you need more "testing" from my side.

<img width="334" height="1095" alt="Image" src="https://github.com/user-attachments/assets/8e03eb11-1037-4cac-80d1-99e6f25e2eda" />

<img width="1369" height="520" alt="Image" src="https://github.com/user-attachments/assets/16be4960-899b-465b-a4b5-841eb022f6fe" />

**Comment by u20p17** [comment]
@marcel151 : thanks for your feedback and good to know that with knx virtual it seems to work fine (y) You are right that we need a KNX router to be able to recieve the milticast messages - an KNX interface is not working for this use case. I will try ro also set up the knx virtual as you described for the future development :)

Regarding the DPTs: Does anybody have expierience which ones normally get used for RGB or RGBW or RGBCCT strips? at the moment we have separte GAs for each color, bit i am not sure if this is the way to go?

**Comment by Alex0176** [comment]
you can install knxd and make a knx router via knxd. 

knxd connect to the knx interface (tunnel mode) and make the router part in it. I used debian and compile knxd by my self. Because the debian 13 knxd packet is broken. 

after you compile it, i think this is the correct paramters in the /etc/knxd.conf 
You find the repository here: https://github.com/knxd/knxd

# configuration for knxd.service /etc/knxd.conf
KNXD_OPTS="-e 0.0.1 -E 0.0.2:8 -u /tmp/eib -R -D -T -S -b ipt:xxx.xxx.xxx.xxx"


# xxx.xxx.xxx.xxx = ip of knx interface. 

# howto compile knxd on debian 13.1
apt install git pkgconf automake  build-essential libtool libsystemd-dev libusb-1.0* cmake libdev-dev
git clone https://github.com/knxd/knxd.git
cd knxd
git checkout main
sh bootstrap.sh
./configure
 make -j$(nproc)
make install
nano /etc/knxd.conf

# debug output of all knx communication
knxtool vbusmonitor1 ip:localhost

```shell
L_Busmon: BC 11 0B 0C 02 D1 00 80 06 :L_Data low from 1.1.11 to 1/4/2 hops: 05 T_Data_Group A_GroupValue_Write (small) 00
L_Busmon: BC 10 FE 02 00 E1 00 80 CE :L_Data low from 1.0.254 to 0/2/0 hops: 06 T_Data_Group A_GroupValue_Write (small) 00
L_Busmon: BC 10 0E 22 00 E3 00 80 0C A6 B6 :L_Data low from 1.0.14 to 4/2/0 hops: 06 T_Data_Group A_GroupValue_Write 0C A6
L_Busmon: BC 10 2E 00 04 E1 00 81 19 :L_Data low from 1.0.46 to 0/0/4 hops: 06 T_Data_Group A_GroupValue_Write (small) 01

```

**Comment by fdepalo** [comment]
my two cents, if you don’t have a knx router and want to simulate a knx network you can use calmiero too, but the learning curve is much steeper…Il giorno 13 set 2025, alle ore 08:06, Alex0176 ***@***.***> ha scritto:﻿Alex0176 left a comment (wled/WLED#2568)
you can install knxd and make a knx router via knxd.
knxd connect to the knx interface (tunnel mode) and make the router part in it. I used debian and compile knxd by my self. Because the debian 13 knxd packet is broken.
after you compile it, i think this is the correct paramters in the /etc/knxd.conf
You find the repository here: https://github.com/knxd/knxd
configuration for knxd.service
KNXD_OPTS="-e 0.0.1 -E 0.0.2:8 -u /tmp/eib -R -D -T -S -b ipt:xxx.xxx.xxx.xxx"
xxx.xxx.xxx.xxx = ip of knx interface.

**Comment by HoffmannOs** [comment]
Now i have it Run, on Bus Monitor in ETS comes all correct on With the PA and Backbone State, in Grey Line 357 is 0.0.1 sending from ETS, is maybe my Problem i have to use these GA 6/0/1 over the Tunnel from The Router from Mainline.

<img width="1553" height="215" alt="Image" src="https://github.com/user-attachments/assets/06413d25-8851-4cf0-b999-3c0801c77b4f" />

**Comment by u20p17** [comment]
Beta9 is out: [WLED_0.15.1_ESP32_KNX.bin](https://github.com/u20p17/WLED/releases/download/v0.15.1_KNXIP_beta9/WLED_0.15.1_ESP32_KNX.bin)

- Added new GA's for H,S,V, HSV, RGB, and RGBW
- Automatically detects LED bus capability and changes GA In/Out accordingly

@HoffmannOs : If i understood correctly it seems that it is now also working in your knx system?

EDIT: Beta 10 now also sends internal temp of ESP and if dallas usermod is loaded also external temp via GA out. (https://github.com/u20p17/WLED/releases/tag/v0.15.1_KNXIP_beta10 )

**Comment by u20p17** [comment]
RC3 is out: https://github.com/u20p17/WLED/releases/tag/v0.15.1-rc.3

- added GAout for Internal temperature + Dallas Temp Sensor (needs add. usermod)
- added GA out for Overtemp. for both sensors
- added GAin for time/date/datetime 
- precompiled versions for all quinled boards



**testers are more than welcome ;)**

@fdepalo: you mentioned that you have designed a KNX-TP hardware for an esp32? can you give me more informations, as this would be for sure another big step for direct communication with KNX :)

**Comment by HoffmannOs** [comment]
Hey,
have flash the Rc7, but have Problems with the State, on Tasmota comes as State back on Bus when turn on or off, but with GA Out Wled Power   [-] (DPT 1.001) nothing comes back on my Routerside, on the Multicast Side shows up, can you maybe change the Point as State to test? Tasmota gives a 1.011 State. Brightness back comes to router as normal...

<img width="1238" height="96" alt="Image" src="https://github.com/user-attachments/assets/ab8b10ee-5321-409d-9870-49247b8ce7d7" />

**Comment by u20p17** [comment]
The actual KNX communication will work the same way (DPT1.001 and DPT1.0.11 are 1-bit boolean values), so you should already see the message on the configured GAout. I will change the labeling in the next RC to make it clearer. I only send the GAouts if they changed - could that be your issue?

**Comment by mwerski** [comment]
Hi @u20p17 , thank you very much for your great work!
I am currently testing with your latest RC8 on a Domraem ESP32 5-channel PWM controller with one white spot connected to GPIO19 and a c/w-white strip on GPIO17/18. WLED UI is working fine and as expected, KNX is basically working but I am experiencing some issues.
One is with segments (but I am aware that this version just implements first steps in regards to segments) but I have another one regarding dimming. In the WLED UI I have 3 possibilities to dim a spot: global brightness, segment brightness and the white level of the segment. The latter one is apparently ignored when sending values via KNX. Another issue is, that there is apparently no transition time when changing brightness levels via KNX, in contrast to changes via WLED UI the brightness changes abrupt.
Other issues - as mentioned - are related to segments, but that can be discussed later ;)
Thx again for this great usermod, I really appreciate it and I hope I can be somewhat helpful in further developments! Just drop me a note when I should do further testings!

**Comment by Alex0176** [comment]
Hi, 
it will be very cool to use the TP too. Like you see it in this link:
https://haus-automatisierung.com/hardware/knx/2019/05/19/knx-und-arduino-selbstbau.html
So we can power up the wled from the knx bus and can control the power supply for the strips

**Comment by abusse** [comment]
Hey everyone,

I wanted to see how feasible and how much work it would be to support TP wire based KNX with this usermod. So I did some testing with this KNX Raspberry Pi HAT: https://hackaday.io/project/171850-knx-raspberry-pi-hat I hooked it up to the serial port of the ESP32. After a little code refactoring to support different transport layers in the usermod, everything worked almost instantly out of the box. So extending the usermod to support both KNX/IP and KNX-TP is not only feasible but relatively straight forward as my prototype shows.

I would gladly drive this implementation forward, but it needs a lot of code restructuring of the usermod as it is now. This, however, should be coordinated as rebasing it would be very painful if substantial changes are made to the usermod as is. What complicates things even more is that I'm currently don't have an KNX/IP gateway at my disposal so I could not test for regressions of the KNX/IP implementation. Therefore my question to @u20p17 , would you be open to cooperate on the code restructuring to have a general interface that can then use different transport layers?

**Comment by fdepalo** [comment]
Hello, sorry for hijacking the thread,

next week i will publish a working usermod for knx-tp, i'm testing it right
now.

Cheers :)

<img width="1275" height="713" alt="Image" src="https://github.com/user-attachments/assets/76e89342-efd2-45fb-86ff-92e1087dd231" />

**Comment by abusse** [comment]
@fdepalo that sounds awesome. I will definitely check it out and test it once available 👍

**Comment by u20p17** [comment]
i am currently working on an OpenKNX application where you can control addressable leds like wled, but with the big advantage that you can programm the device via ETS (TP and IP will be available)… still work to do but the first alpha version is already running at different OpenKNX developers ;)

**Comment by fdepalo** [comment]
my personal opinion: ets is not an advantage but an handicap ;) better to program in yaml Il giorno 17 gen 2026, alle ore 22:25, u20p17 ***@***.***> ha scritto:﻿u20p17 left a comment (wled/WLED#2568)
i am currently working on an OpenKNX application where you can control addressable leds like wled, but with the big advantage that you can programm the device via ETS (TP and IP will be available)… still work to do but the first alpha version is already running at different OpenKNX developers ;)
